src/Microsoft.OData.Edm/Schema/EdmTargetPath.cs

```diff
@@ -0,0 +1,111 @@
+//---------------------------------------------------------------------
+// <copyright file="EdmTargetPath.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.OData.Edm
+{
+    /// <inheritdoc/>
+    public class EdmTargetPath : IEdmTargetPath
+    {
+        private List<IEdmElement> segments;
+        private readonly string path;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EdmTargetPath"/> class.
+        /// </summary>
+        /// <param name="segments">Target path segments.</param>
+        public EdmTargetPath(IEnumerable<IEdmElement> segments)
+        {
+            EdmUtil.CheckArgumentNull(segments, nameof(segments));
+
+            this.segments = segments.ToList();
+
+            ValidateSegments();
+
+            this.path = this.GetPathString();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EdmTargetPath"/> class.
+        /// </summary>
+        /// <param name="segments">Target path segments.</param>
+        public EdmTargetPath(params IEdmElement[] segments)
+            :this(segments.AsEnumerable())
+        {
+        }
+
+        /// <inheritdoc/>
+        public IReadOnlyList<IEdmElement> Segments
+        {
+            get { return this.segments; }
+        }
+
+        /// <inheritdoc/>
+        public string Path
+        {
+            get { return this.path; }
+        }
+
+        /// <summary>
+        /// Checks if this target path is equal to another target path.
+        /// </summary>
+        /// <param name="other">The other target path to compare it to.</param>
+        /// <returns>True if the two target paths are equal.</returns>
+        /// <exception cref="ArgumentNullException">Throws if the input other is null.</exception>
+        internal bool Equals(EdmTargetPath other)
+        {
+            EdmUtil.CheckArgumentNull(other, nameof(other));
+
+            if (this.segments.Count != other.segments.Count)
+            {
+                return false;
+            }
+
+            for (int index = 0; index < this.segments.Count; index++)
+            {
+                if (!this.segments[index].Equals(other.segments[index]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void ValidateSegments()
+        {
+            // Validate no segment is null
+            if (this.segments.Contains(null))
+            {
+                throw Error.ArgumentNull(nameof(segments));

```

Comment:
The error messages that would be result from this statement would be confusing/misleading since it'd be pointing out that `segments` is null yet it's one or more elements in the `segments` collection that is null.

Reply:
Same thing here https://github.com/OData/odata.net/blob/d1f9201e4b0d72e76c1c7d99d1b86b7af88e29bb/src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs#L41
Not sure of the correct exception/message to throw